src/e2ir.c

```diff
@@ -5174,27 +5175,78 @@ elem *toElem(Expression *e, IRState *irs)
 
             elem *e = NULL;
 
+            size_t dim = sle->elements ? sle->elements->dim : 0;
+            assert(dim <= sle->sd->fields.dim);
+
             if (sle->fillHoles)
             {
                 /* Initialize all alignment 'holes' to zero.
                  * Do before initializing fields, as the hole filling process
                  * can spill over into the fields.
-                 *
-                 * TODO: Currently any fields are conservatively filled to zero,
-                 * even if a field will be set immediately after.
                  */
+                const size_t structsize = sle->sd->structsize;
                 size_t offset = 0;
-                for (size_t i = 0; i < sle->sd->fields.dim; i++)
+                //printf("-- %s - fillHoles, structsize = %d\n", sle->toChars(), structsize);
+                for (size_t i = 0; i < sle->sd->fields.dim && offset < structsize; )
                 {
                     VarDeclaration *v = sle->sd->fields[i];
-                    size_t vend = v->offset + v->type->size();
-                    e = el_combine(e, fillHole(stmp, &offset, vend, sle->sd->structsize));
+                    if (i < dim && (*sle->elements)[i] && v->type->size())
+                    {
+                        //if (offset != v->offset) printf("  1 fillHole, %d .. %d\n", offset, v->offset);
+                        e = el_combine(e, fillHole(stmp, &offset, v->offset, structsize));
```

Comment:
Yes, and built-in `tupleof` and others depend on the order of `fields`. Changing it will be problematic.